core/trino-main/src/main/java/io/trino/operator/aggregation/ApproximateSetVarcharAggregation.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.aggregation;
+
+import io.airlift.slice.Slice;
+import io.airlift.stats.cardinality.HyperLogLog;
+import io.trino.operator.aggregation.state.HyperLogLogState;
+import io.trino.spi.block.BlockBuilder;
+import io.trino.spi.function.AggregationFunction;
+import io.trino.spi.function.AggregationState;
+import io.trino.spi.function.CombineFunction;
+import io.trino.spi.function.InputFunction;
+import io.trino.spi.function.LiteralParameters;
+import io.trino.spi.function.OutputFunction;
+import io.trino.spi.function.SqlType;
+import io.trino.spi.type.StandardTypes;
+
+import static io.trino.operator.aggregation.ApproximateSetAggregationUtils.getOrCreateHyperLogLog;
+
+/**
+ * // TODO(https://github.com/trinodb/trino/issues/8346)
+ * approx_set for VARCHAR(x) needed to be split out from ApproximateSetGenericAggregation because of limitations of
+ * annotation based mechanism for defining aggregation functions.
+ * It was not possible to have @TypeParameter("T") for some input-function overrides and @LiteralParameters("x") for other.
+ */
+@AggregationFunction("approx_set")
+public final class ApproximateSetVarcharAggregation
+{
+    private ApproximateSetVarcharAggregation() {}
+
+    @InputFunction
+    @LiteralParameters("x")
+    public static void input(
+            @AggregationState HyperLogLogState state,
+            @SqlType("varchar(x)") Slice value)
```

Comment:
could this be handled with `@SqlType("T")` and so without need for `@LiteralParameters("x")` and separate class?

Reply:
I think I tried that. But let me double check

Reply:
Oh I know.

 I could drop that altogether. And varchar(x) would be handled by
```java
    @InputFunction
    @TypeParameter("T")
    public static void input(
            @OperatorDependency(
                    operator = XX_HASH_64,
                    argumentTypes = "T",
                    convention = @Convention(arguments = NEVER_NULL, result = FAIL_ON_NULL))
                    MethodHandle methodHandle,
            @AggregationState HyperLogLogState state,
            @SqlType("T") Object value)
    {
```
but then the generated HLL would be different as override for varchar uses  `Murmur3Hash128.hash64(value)` while version with `@TypeParameter(T)` uses `XX_HASH_64` operator which translates to `XxHash64.hash(value)`

I wanted to keep behaviour unchanged for sake of users which store generated HLL in some external storage for future use.

Reply:
> I wanted to keep behaviour unchanged for sake of users which store generated HLL in some external storage for future use.

i agree.

We could actually have a stability test for `approx_set` to reinforce persistable guarantees from docs.

> @SqlType("T") Object value

Would `@SqlType("T")  Slice value` overload work?

but then, it would not necessarily be correct.
For example, two equal values of some Slice-backed type other than varchar, would hash differently, if we skip type-provided hash function

Reply:
> > Would @SqlType("T") Slice value overload work?

> but then, it would not necessarily be correct.
> For example, two equal values of some Slice-backed type other than varchar, would hash differently, if we skip type-provided hash function

It would work I guess. And yeah - theoretically it could not be correct. Not sure if have such hashfunction though.